incremental-font-transfer/src/uri_templates.rs

```diff
@@ -0,0 +1,682 @@
+//! Implementation of the specific variant of URI template expansion required by the IFT specification.
+//!
+//! Context: <https://w3c.github.io/IFT/Overview.html#uri-templates>
+//!
+//! In IFT RFC6570 style uri templates are used, however the IFT specification restricts template syntax
+//! to a subset (level 1 with a predefined set of variables) of the full RFC6570 syntax. This implements
+//! a URI template expander that adheres to the IFT specific requirements.
+//!
+//! By implementing our own we avoid pulling in a much larger general purpose template expansion library
+//! and improve performance versus a more general implementation.
+use std::{ops::RangeInclusive, sync::OnceLock};
+
+enum ParseState {
+    // Literal parsing
+    Literal,
+    LiteralPercentEncoded(Digit),
+
+    // Expression parsing,
+    Expression(Variable),
+}
+
+/// Represents the process of matching one of the predefined variable names: id, id64, d1, d2, d3, d4
+enum Variable {
+    Begin,
+    I,
+    ID,
+    ID6,
+    ID64,
+    D,
+    DX(u8),
+    Undefined,
+    Dot,
+    PercentEncoding(Digit),
+}
+
+/// Which digit of a percent encoding we're on
+enum Digit {
+    One,
+    Two,
+}
+
+#[derive(Default)]
+struct OutputBuffer(String);
+
+/// Indicates a malformed URI template was encountered.
+///
+/// More info: <https://datatracker.ietf.org/doc/html/rfc6570#section-3>
+#[derive(Debug, PartialEq, Eq)]
+pub struct UriTemplateError; // TODO(garretrieger): change patchmap to use this.
+
+impl std::fmt::Display for UriTemplateError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "Invalid URI template encountered.")
+    }
+}
+
+impl std::error::Error for UriTemplateError {}
+
+/// Implements uri template expansion from incremental font transfer.
+///
+/// Specification: <https://w3c.github.io/IFT/Overview.html#uri-templates>
+///
+/// IFT uri templates are a subset of the more general <https://datatracker.ietf.org/doc/html/rfc6570>
+/// uri templates. Notably, only level one substitution expressions are supported and there are a fixed
+/// set of variables used in the expansion (id, id64, d1, d2, d3, and d4).
+///
+/// All arguments are assumed to be utf8 encoded strings.
+pub(crate) fn expand_template(
+    template_string: &str,
```

Comment:
would it make more sense to pass this in as bytes?

Reply:
It's important that we're operating on the bytes of a utf8 encoding, and it's my understanding that being a &str implies the contents are utf8 encoded.